### ISSUE TITLE
Pre-training issue

### ISSUE THREAD

**Issue by guangyuyang001** [YunshiWen/VQShape#4]
I followed the pre-training method you provided, but the model did not start training and output the following. I have not found a suitable solution. Could you please provide me with the steps to solve it?

Datasets: TRAIN 0 | VAL 0

Loading `train_dataloader` to estimate number of stepping batches.

  | Name  | Type    | Params | Mode 
------------------------------------------
0 | model | VQShape | 37.1 M | train
------------------------------------------
37.1 M    Trainable params
0         Non-trainable params
37.1 M    Total params
148.339   Total estimated model params size (MB)
184       Modules in train mode
0         Modules in eval mode
`Trainer.fit` stopped: No training batches.

**Comment by guangyuyang001** [comment]
I've found the problem, which is that "dataset_dirs.extend(glob.glob(f"{data_root}/{task}/{split}/\*"))" in the TimeSeriesDatasetLazy.py file should leave out the "/\*". **I don't know if this is right or not?** If you don't think it should be removed then there is something wrong with your preprocessor, because the next level in the uea data set from "data/uea/Train" and "data/uea/Test" is ".csv".

**Comment by YunshiWen** [comment]
There seems to be a mismatch between the data loading script and the preprocessing notebook...

The dataset folder hireachy should be `data_root/uea/split/dataset_name/sample_id.csv`. The notebook has been updated to fix this. Thanks for pointing out this issue! And please let me know if the issue is addressed.

**Comment by guangyuyang001** [comment]
Thank you for quick response! I just checked the update notebook, and the issue seems to be resolve now. Everything is working as expected. Appreciate your help!
I noticed that the configuration in the pretrain.sh script seems to be different from the one mentioned in the paper "Model Configurations." Which one should this be in order to reproduce the results given in the paper?

**Comment by YunshiWen** [comment]
I believe the confusion is related to the number of encoder layers. In short, please use the configuration in `pretrain.sh`.

When the paper was initially written, we use a 8-layer PatchTST as the encoder, which is reported in the paper. However, in later developments, to consider other tasks beyond classification, we modify the encoder structure by using two modules: an encoder backbone and a tokenizer, each with 4 layers. This design is more flexible as the encoder backbone can be replaced by other architectures (such as ResNet) and it decouples the number of patches with the number of shape tokens. Also, since both the encoder and the tokenizer are Transformer, the new design is pretty much equivalent to the architecture used in the paper. In experiments, we also find that these two designs result in approximately the same performance, despite they have different number of parameters.

Since we were asked not to introduce "significant" changes to the revised paper, we did not detail this change to avoid further confusion and additional changes. All the results reported in the paper and the checkpoints are obtained using the new design.

**Comment by guangyuyang001** [comment]
This really clarifies my confusion. Thank you again for your detailed response! 
I have another question about the implementation of the tokenize method in the VQShape class (vqshape/model.py). It seems that the architecture described in the paper (patch-based Transformer encoder, Time-series Encoder, and Attribute Decoder) differs from the code implementation, which includes a patch-based Transformer encoder, a Time-series Encoder, self.tokenizer (a TransformerDecoder), and another Time-series Encoder. I couldn’t find a mention of self.tokenizer in the paper—did I misunderstand its role, or is this a later design change not reflected in the paper?

**Comment by YunshiWen** [comment]
The mapping from the modules in the paper and the modules in the code are as the follow:
- TS encoder: `self.encoder` and `self.tokenizer`,
- TS decoder: `self.decoder`,
- Shape decoder: `self.shape_decoder`,
- Attribute decoder: `self.attr_decoder`,
- Attribute encoder: `self.attr_encoder`.

The 4-layer `self.encoder` and the 4-layer `self.tokenizer` together are equivalent to the 8-layer encoder described in the paper, which is the new designed mentioned above.

The tokenizer maps the feature embeddings from the encoder to the embeddings of the shape tokens which is then decoded by the attribute decoder. The tokenizer takes a set of learned prompt tokens and extract information from the features embeddings (the encoder's outputs) through cross-attention.

The tokenizer is implemented using the `nn.TransformerDecoder` class to use the cross-attention mechanism, but it always have the full attention. In other words, it is just an encoder with self-attention and cross-attention.
